File: Examples/Ex12/Ex12.cpp
/************************************************************************
	
	Copyright 2012-2013 Luciano Buglioni
 
	Contact: luciano.buglioni@gmail.com

	This file is a part of FluxSol

	FluxSol is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    any later version.

    Free CFD is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    For a copy of the GNU General Public License,
    see <http://www.gnu.org/licenses/>.

*************************************************************************/

///////////////
// EXAMPLE 1 //
///////////////
// Viwcoplastic Problem
// 1 Elements,4 node each, loaded as follows
//FORCEX
// ---------------
// |             |             
// |             |             
// |      1      |           
// |             |             
// |             |             
// ---------------
// FIX_XY		FIX_Y

//Solved with Laspack Library

#include "EPSol.h"

//Sparse libraries
//#include "laspack.h"

#include <iostream>
#include <vector>

using namespace std;
using namespace FluxSol;

class Eulerian_ViscoPlastic{

public:
    Eulerian_ViscoPlastic();
    void setup();
	void assemble();
	void calc_res();

private:
//Matrices
//Gauss
// ELEMENTAL MATRICES AND VECTORS
    GaussFullMatrices B,J,dHrs;
    //Shape
    Matrix<double> Nv,NsigF,Ns;//Nv,Nsig,NF,Ns;
    
	//Shape derivatives
	Matrix<double> Bv,Bs,LM;
	vector < Matrix <double> >BsigF, BL;
	Matrix <double> temp4x16;

	FluxSol::Matrix<double> G,H,Kel;
	
	//Vectors
	//Solution
	Matrix<double> U,Uv,Usig,UF,Us;
	//Deformation gradients
	Matrix<double> F,P,L;	//Unsymmetric vectors
	
	//Residual 
	vector < vector <Matrix <double> > > Kt; 

	Matrix<double> R;	//Nv,Nsig,NF,Ns;
	
	//Dimensions
	int dof[4];
	enum field;
	
	//Material
	Matrix<double>c;
	double Ey,nu;

	Matrix<double> Ve;

    ofstream logfile;
};


void Eulerian_ViscoPlastic::setup()
{

	ofstream logfile;
	logfile.open("Logfile.txt");
	
	//Matrices
	//Constants should be defined here
	G=H=Matrix<double>(4,4);
	G[0][0]=G[1][3]=G[2][3]=G[2][1]=1.;
	H[0][0]=G[0][1]=1.;
	
	Ey = 200.0e9;
	nu = 0.33;
	
	//Material
    c=Matrix<double>(4,4);
	double ck=Ey/((1.+nu)*(1.-2.*nu));
	c[0][0]=c[1][1]=c[2][2]=ck*(1.-nu);
	c[0][1]=c[0][2]=c[1][0]=c[2][0]=c[1][2]=c[2][1]=ck*(1.-nu);
	c[2][2]=ck*(0.5-nu);
	
	const FluxSol::FEIntegrationScheme intsch(1,2);


	//FluxSol::QuadLinearElement e(v);
	FluxSol::Element<2> e(v);
	e.Set_Nodes(4, 0, 1, 2, 3);
	//logfile << "Element Gauss Order: " << e.GaussOrder() << "\n\n";
	
	//Shape functions 
	//Nv = e.H();
	Bv=Matrix<double>(4,8);

	std::vector< FluxSol::Element<2> > ve;
	ve.push_back(e);
	FluxSol::FeGrid <2> g(v, ve);
    FluxSol::FEValues<2> fev(e, g);
    
	J = fev.Jacobian();
    B = fev.shape_grad_matrix();
   
    dHdrs=fev.shape_localgrad_matrix(); //COMPLETE
    ShapeFunctionGroup shfngr = e.CreateShapeFunctionGroup();
	// Shape Functions
	//To MODIFY REFER TO dim
	//Shape and gradient MATRICES
	//These are not grouped into an array because
	//1: Two of them are equal
	//2: Bsig and BF are THREE DIM ARRAY
    Nv=Matrix<double>(2, 8);//v
	NsigF[1]=Matrix<double>(4,16);//sig
	N[3]=Matrix<double>(1, 4);//s

	//Solution (Fig 2.1)
	dof[0]=8;		//Velocity: dim=2 x 4 nodes
	dof[1]=[2]=16;	//Deformation
	dof[3]=4;
	//R   =Matrix<double>(11,1);
	for (int i=0;i<4;i++)
		R[i]=Matrix<double>(dof[i],1);
	//Deformation gradients
	//Vector<double> F,P,L;	//Unsymmetric vectors
	
	// #Nodal values
	// Ve=Matrix<double>(8, 1)))


	// #Derivatives
	// dHxy=Matrix<double>(2, 4)))
	// Bs=Matrix<double>(2, 8)))
	// Bv=Matrix<double>(4, 8)))
	// #BsigF=[Matrix<double>(4, 16))),Matrix<double>(4, 8)))]
	// BsigF=arange(128).reshape(4,16,2) #
	// temp4x16=Matrix<double>(4, 16)))
	// B4i=arange(32).reshape(4,4,2) #
	vector < vector <vector double > > B4i;
	// #(4,16,2)
	// print(BsigF[0])
	for (int i=0;i<2;i++)	BsigF.push_back(<Matrix<double>(4, 16));
	temp4x16=<Matrix<double>(4, 16)
	
	LM=Matrix<double>(4, 4)

	//R =Matrix<double>(44, 1)
	RF  =Matrix<double>(16, 1)
	Rsig=Matrix<double>(16, 1)
	Rs  =Matrix<double>(4, 1)
	Rv  =Matrix<double>(8, 1)

	// U =Matrix<double>(44, 1)))
	UF  =Matrix<double>(16, 1);
	Usig=Matrix<double>(16, 1);
	dVxy=Matrix<double>(4, 2);

	// #Symmetric tensors
	E=Matrix<double>(4, 1)))

	// #Stress
	sig=Matrix<double>(4, 1);   #Stress Gauss Points
	sig_d=Matrix<double>(4, 1); #Deviatoric

	P=Matrix<double>(4, 1);


	v=Matrix<double>(2, 1);    #Gauss Point velocity			


	Kel=Matrix<double> (44, 44);
	
	// for (int i=0;i<4;i++)
		// for (int j=0;j<4;j++)
			// for (int k=0;k<4;k++)
				// B4i.push_back(0.)
			
	for (int i=0;i<4;i++)
		BL.push_back(Matrix<double>(4,2));
							
}


void Eulerian_ViscoPlastic::assemble()
{

	for (int i=0;i<4;i++){
	//C = E / (1 - nu*nu)*[1 nu 0; nu 1 0; 0 0 (1 - nu) / 2];
	cout << "Num Integration Points"<<intsch.NumPoints()<<endl;
	for (int g = 0; g < intsch.NumPoints(); g++)
	{

		Bs=fev.shape_grad_comps(g);
		 
		Bv[0][2*i  ]=B[2][2*i]=Bs[0][i];
		Bv[1][2*i+1]=B[2][2*i]=Bs[0][i];


		for (int k=0;k<4;k++){
		#shape functions
		Nv[0][2*k ]=Nv[1][2*k+1]=Ns[0,k];
		#derivatives Bv (B.14)
		Bv[0][2*k  ]=dHxy[0,k];
		Bv[1][2*k  ]=dHxy[1,k];
		Bv[2][2*k+1]=dHxy[0,k];
		Bv[3][2*k+1]=dHxy[1,k];}
		
		for (int k=0;k<8;k++){
			BL[1][1][k]=BL[3][3][k]=Bv[1][k];
			BL[1][2][k]=BL[3][4][k]=Bv[3][k];
			BL[2][1][k]=BL[4][3][k]=Bv[2][k];
			BL[2][2][k]=BL[4][4][k]=Bv[4][k];
			BL[1][3][k]=BL[1][4][k]=BL[2][3][k]=BL[2][4][k]=BL[3][1][k]=BL[3][2][k]=BL[4][1][k]=BL[4][2][k]=0.;}
		
		for (int i=0;i<4;i++){
			for (int l=0;l<4;l++)
				for (int m=0;m<4;m++)
					for (int n=0;n<4;n++)
						if (l==m)
							B4i[l,m,n]=Bs[n,i];
						else
							B4i[l,m,n]=0.;        
			for (int l=0;l<4;l++)
				for for (int m=0;m<4;m++) 
					for (int n=0;n<4;n++)
						BsigF[4*i+l][m][n]=B4i[l][m][n];}//For i 
		
			
		//FluxSol::Matrix<double> Kg = B.Mat(g).Tr()*c*B.Mat(g);
		for (int r = 0; r < 8; r++)
			for (int c = 0; c < 8; c++){
				Kel[r][c] += Kg[r][c] * intsch[g].w()*J.Mat(g).det();
				cout << "weight" << intsch[g].w()<<endl;
				cout << "Kel: "<< Kel[r][c]<<endl;	
	}}}


	logfile << "Stiffness Matrix \n\n";
	logfile << Kel.outstr();

	//DOFs 5, 6 and 8 restricted
	for (int i = 0; i < 8; i++)
	{
		Kel[4][i] = Kel[i][4] = 0.0;
		Kel[5][i] = Kel[i][5] = 0.0;
		Kel[7][i] = Kel[i][7] = 0.0;
	}
	Kel[4][4] = Kel[5][5] = Kel[7][7] = 1.;
	// //////////////////// LASPACK SOLVER ///////////////////////

	PETSC_Solver<double,2> solver(8);

	
	//DOFs 5, 6 and 8 restricted
	for (int i = 0; i < 8; i++)
	{
		Kel[4][i] = Kel[i][4] = 0.0;
		Kel[5][i] = Kel[i][5] = 0.0;
		Kel[7][i] = Kel[i][7] = 0.0;
	}
	Kel[4][4] = Kel[5][5] = Kel[7][7] = 1.;
	
	for (int row = 0; row < 8; row++){
		for (int col = 0; col < 8; col++)
		{
			solver.AddMatVal(row,col,Kel[row][col]);
		}

	}
	cout<<Kel.outstr();
	solver.Flush();
	
	solver.SetbValues(2,1000.0);
	

	solver.ViewInfo();
	solver.Solve();
	solver.ViewInfo();
	
	logfile.close();
	return 0;


}

void Eulerian_ViscoPlastic::calc_res()
{
	//R[0]=B[0].tr()*P - N[0].tr()*tP;
	// Rsig j = Bsig mjk * vk * Usig j - C mj E.(e)j   
	//R[1]=(N[1].tr()+tau(e)*v) * ()
	
}

File: Examples/Ex12/ex12.py
# https://docs.sympy.org/latest/tutorial/matrices.html
#Plain strain/stress 1 element example
from numpy import *
import numpy.matlib #for zeros

import array as arr

#X is combines per row to calculate J
p=1.0/1.732050807568877
gauss=[-p,p]
wi=[0.777,0.777]
#Numerated as in Bathe
X2=matrix([[1,1],[0,1],[0,0],[1,0]])
#Numerated as in deal.ii
#X2=matrix([[0,0],[1,0],[0,1],[1,1]])

#Nodal values
Ve=matrix(numpy.matlib.zeros((8, 1)))

#Shape Functions
Ns=matrix(numpy.matlib.zeros((1, 4)))
Nv=matrix(numpy.matlib.zeros((2, 8)))
NsigF=matrix(numpy.matlib.zeros((4, 16)))

#Derivatives
dHxy=matrix(numpy.matlib.zeros((2, 4)))
Bs=matrix(numpy.matlib.zeros((2, 8)))
Bv=matrix(numpy.matlib.zeros((4, 8)))
#BsigF=[matrix(numpy.matlib.zeros((4, 16))),matrix(numpy.matlib.zeros((4, 8)))]
BsigF=arange(128).reshape(4,16,2) #
temp4x16=matrix(numpy.matlib.zeros((4, 16)))
B4i=arange(32).reshape(4,4,2) #
#(4,16,2)
print(BsigF[0])
LM=matrix(numpy.matlib.zeros((4, 4)))

K=matrix(numpy.matlib.zeros((44, 44)))
R=matrix(numpy.matlib.zeros((44, 1)))
dVxy=matrix(numpy.matlib.zeros((4, 2)))

#Symmetric tensors
E=matrix(numpy.matlib.zeros((4, 1)))
sig_d=matrix(numpy.matlib.zeros((4, 1))) #Deviatoric
sig=matrix(numpy.matlib.zeros((4, 1)))


#Material properties (Table 2.1 p28)
#HSLA-65 steel with strain rate between e-3 and e-4
A0=6.34e11
psi=3.25
m=0.1956
n=0.06869
s0=80.0e6


print (K)
print (X2[0,0])
print (X2[2,1])

ey=200.e9
nu=0.33

ck = ey/ ((1. + nu)*(1. - 2. * nu))
c=matrix(numpy.matlib.zeros((4, 4)))
c[0,0]=c[1,1]=c[2,2]=ck*(1.-nu)					
c[0,1]=c[1,0]=c[0,2]=c[2,0]=c[2,1]=c[1,2]=ck*nu
c[3,3]=ck*(.5 - nu)
print ("C matrix")
print(c)
#Set G and H matrices
#2.32
G=matrix([[1,0,0,0],[0,0,0,1],[0,0,0,1],[0,1,0,0]]) 
H=matrix([[1,0,0,0],[0,1,0,0],[0,0,0,0],[0,0,0,0]]) 
    
for e in range (4):
    #Obtain Ve from global
    for ig in range(2):
        for jg in range(2):
            r=gauss[ig]
            s=gauss[jg]

            #Numerated as in Bathe
            Ns  =0.25*matrix([(1+s)*(1+r),(1-r)*(1+s),(1-s)*(1-r),(1-s)*(1-r)])            
            dHrs=matrix([[(1+s),-(1+s),-(1-s),(1-s)], [(1+r),(1-r),-(1-r),-(1+r)] ])
            #Numerated as in deal.ii
            #dHrs=matrix([[-(1-s),(1-s),-(1+s),(1+s)], [-(1-r),-(1+r),(1-r),(1+r)] ])        
            dHrs/=4
            J=dHrs*X2
            dHxy=linalg.inv(J)*dHrs
            detJ=linalg.det(J)
            #Calculate shape functions
            #Bs=J-1 dHrs(B.13)
            Bs=dHxy
            for k in range(4):
                #shape functions
                Nv[0,2*k  ]=Nv[1,2*k+1]=Ns[0,k]
                #derivatives Bv (B.14)
                Bv[0,2*k  ]=dHxy[0,k]
                Bv[1,2*k  ]=dHxy[1,k]
                Bv[2,2*k+1]=dHxy[0,k]
                Bv[3,2*k+1]=dHxy[1,k]
                
            for i in range(4):
                for l in range(4):
                    for m in range(4):  
                        for n in range(2):
                            if (l==m):
                                B4i[l,m,n]=Bs[n,i]
                            else
                                B4i[l,m,n]=0.
                BsigF[i,]=B4i[l,m,n]
            
            #Galerkin strain integration
            #Calculate deformation gradient Fij, for that
            #Calculate Velocity gradient Lij
            #Lij=dvi/dxj(2.4) 
            #According to B.11 d(phi)/dxj=J-1(ij) dphi/drj = Bvjk Vk
            dVxy=Bv*Ve #(4x8)*(8x1)=4x1(4x1)(vx,x vx,y vy,x vy,y)T 
            
            #Stabilization factor tau 2.26
            #tau=beta*he/(2|v|)
            #See beta estability paramter
            #LM (2.33 p23)
            #Attention: SE DIFFERENCES WITH L_ in 2.28
            LM[0,0]=LM[2,2]=dVxy[0]
            LM[0,1]=LM[2,3]=dVxy[1]
            LM[1,0]=LM[3,2]=dVxy[2]
            LM[1,1]=LM[3,3]=dVxy[3]
            
            #BL interpolators BLijk (4,4,8) (B.17 p165)
            
            #Set tractions tP (2.34)
            
            
            #Before Strain tensors, DEFORMATIONS
            #With evp = f(sigma,s) (Sec. 2.4.2)
            #eps_vp=A sinh (psi (sigma/s))^(1/m) (2.57 p27)
            #Calculate Rate of Def Tensor D (2.13, 2.14)
            #D(th)ij=alpha vk dT/dxk deltaij 2.13
            #IN THIS EXAMPLE THIS DTH 0
            #D(vp)ij=sqrt(3/2) e. vp Nij  2.14
            #Nij Direction of plastic flow
            #
            #Calculate Leij = Lij - D(th)ij - D(vp) ij (2.10-2.12)
            #
            #Calculate La (ij)
            #Laij=F(-1)ki F(-1)kl Le lj
            #Calculate Almansi deformation gradient E (A.5)
            #Ea ij= 1/2(Lki F(-1)lk F(-1)LJ +F(-1)ki F(-1)KL Llj )

            w=0.25 #TO MODIFY
            #Calculate sigma
            #2.31 Comes first from 2.2
            #Pij=vk d(sig ij)/xk - dvi/dxk sig(kj) + dvk/dxk sig ij
            #Calculate Piola Kirchoff Pi (2.31) Gij Cjk˙¯-Gij LM (jk) sig(k) + 
            #Attention double contraction
            P=G*c*E-G*LM*sig+(LM[0,0]+LM[1,1])*G*sig
            
            #RESIDUALS ******************* 2.36 to 2.39 *****************************
            Rv  =Bv.transpose()*P*w*detJ #Remains the summ of particular gauss points
            #Rsig[16x1] (4 per node)
            #Construct vk Bsig mik
            #for k in range(4):
            #    temp=temp+
            Rsig=(NsigF+NsigF).transpose()
            #print (B)
            #K+=(B.transpose()*c*B*w)
            #print (K)
    #print (K)

#Boundary conditions
#Numerated as in Bathe
# for i in range(8):
    # K[4,i] = K[i,4] = 0.0
    # K[5,i] = K[i,5] = 0.0
    # K[7,i] = K[i,7] = 0.0
	
# K[4,4] = K[5,5] = K[7,7] = 1.;
# R=[0,0,1000.0,0,0,0,0,0]

# #Numerated as in deal.ii
# for i in range(8):
    # K[0,i] = K[i,0] = 0.0
    # K[1,i] = K[i,1] = 0.0
    # K[3,i] = K[i,3] = 0.0
	
# K[0,0] = K[1,1] = K[3,3] = 1.;
# R=[0,0,0,0,1000.0,0,0,0]

print (K)

#U=linalg.solve(K, R)
print ("Results")
#print(U)


